src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -225,48 +177,183 @@ void BaseCouplingScheme::finalize()
+}
+
+void BaseCouplingScheme::initializeData()
+{
+  // InitializeData uses the template method pattern (https://en.wikipedia.org/wiki/Template_method_pattern).
+  PRECICE_CHECK(_isInitialized, "initializeData() can be called after initialize() only.");
+  PRECICE_CHECK(not _initializeDataHasBeenCalled, "initializeData() can only be called once.");
+  _initializeDataHasBeenCalled = true;
+  PRECICE_TRACE("initializeData()");
+
+  if (not _sendsInitializedData && not _receivesInitializedData) {
+    PRECICE_INFO("initializeData is skipped since no data has to be initialized.");
+    return;
+  }
+
+  PRECICE_DEBUG("Initializing Data ...");
+
+  PRECICE_CHECK(not(_sendsInitializedData && isActionRequired(constants::actionWriteInitialData())),
+                "InitialData has to be written to preCICE before calling initializeData().");
+
+  _hasDataBeenExchanged = false;
+
+  exchangeInitialData();
+}
+
+void BaseCouplingScheme::advance()
+{
+  PRECICE_TRACE(_timeWindows, _time);
+  checkCompletenessRequiredActions();
+  PRECICE_ASSERT(_isInitialized, "Before calling advance() coupling scheme has to be initialized via initialize().");
+  PRECICE_CHECK((not _receivesInitializedData && not _sendsInitializedData) || (_initializeDataHasBeenCalled),
+                "initializeData() needs to be called before advance if data has to be initialized.");
+  _hasDataBeenExchanged = false;
+  _isTimeWindowComplete = false;
+
+#ifndef NDEBUG
+  for (const DataMap::value_type &pair : getReceiveData()) {
+    Eigen::VectorXd &  values = *pair.second->values;
+    int                max    = values.size();
+    std::ostringstream stream;
+    for (int i = 0; (i < max) && (i < 10); i++) {
+      stream << values[i] << " ";
+    }
+    PRECICE_DEBUG("Begin advance, first New Values: " << stream.str());
+  }
+#endif
+  PRECICE_ASSERT(_couplingMode != Undefined);
+
+  if (reachedEndOfTimeWindow()) {
+
+    _timeWindows += 1;  // increment window counter. If not converged, will be decremented again later.
+
+    std::pair<bool, bool> convergenceInformation = exchangeDataAndAccelerate();
+
+    if(isImplicitCouplingScheme()) {  // check convergence
+      bool convergence = convergenceInformation.first;
+      bool convergenceCoarseOptimization = convergenceInformation.second;
+      if (not convergence) {  // repeat window
+        PRECICE_DEBUG("No convergence achieved");
+        requireAction(constants::actionReadIterationCheckpoint());
+        // The computed time window part equals the time window size, since the
+        // time window remainder is zero. Subtract the time window size and do another
+        // coupling iteration.
+        PRECICE_ASSERT(math::greater(_computedTimeWindowPart, 0.0));
+        _time = _time - _computedTimeWindowPart;
+        _timeWindows -= 1;
+      } else {  // write output, prepare for next window
+        PRECICE_DEBUG("Convergence achieved");
+        advanceTXTWriters();
+        PRECICE_TRACE(_timeWindows, _time);
```

Comment:
`PRECICE_TRACE` is only allowed when entering a function.

Reply:
I removed this line and the one below in 792a6468e168a99e7fc562531d427b323cf356ad